### DISCUSSION TITLE
Tooling for automatically refactoring MachineConfig into a serializable form

### DISCUSSION THREAD

**Discussion by awreccan** [statelyai/xstate#2160] [Q&A]
Is there tooling like the following available? (made by xstate creators and/or the community)

```js
const [machineConfig: MachineConfig, options: MachineOptions] = tooling(machineConfig: MachineConfig)
```

Essentially something that can walk a MachineConfig and collect any functions it encounters into a MachineOptions object, and replacing them with corresponding strings in the MachineConfig, such that the machineConfig output is serializable.

It would require intimate awareness of the various types of each field, so I'm a little afraid of biting the bullet and implementing it on my own.

EDIT: I see a similar request here: https://github.com/davidkpiano/xstate/discussions/2064

**Comment by Andarist** [comment]
Basically, all implementation details can be externalized in the `options`. If you need the config serializability then why do you put function there instead of putting them in the `options` and referencing them with strings in the `config`?

**Reply by awreccan** [reply]
Tldr: xstate has great composability built into its API, and that enables scaling the writing of a machine config across multiple devs, sharing states used across multiple machines. But the serialization feature is not scalable as of yet - the manual work needed is a blocker for working at higher levels of abstraction.

I'm writing a library for developers to build upon. The library hides the implementation details of lots of xstate nested states inside. The focus is on developer experience. The library and the framework it's creating is leveraging on the composability of an xstate machine config - for example, one dev could be stitching together a StateNodeConfig made up of:

- 1 atomic state, newly created with a helper method that enforces some framework invariants (some reserved state names, some nesting structure violations, etc), 
- 2 reused preexisting states (compound or atomic), created by other devs for other xstate machines in our codebase (maybe modify 1 of those 2 in very small ways) - (fyi some of these reusable compound state configs can be pretty complicated, so reuse is the way to go to prevent code bloat, to prevent bugs, and to encapsulate a single source of truth for a certain compound state's internal logic)
- etc 

The final StateNodeConfig is passed to a **transformer** that nests it inside a compound state with some other top-level states added and some transitions added to all the dev-provided states. 
 
So the only time in this dev process when it's possible to externalize options to achieve serializability is once the full machine config is available, and that is inside the final library-provided **transformer**. It can't be done by individual devs contributing to individual states.
